autopilot/accounts.go

```diff
@@ -0,0 +1,187 @@
+	"go.sia.tech/core/types"
+	"go.sia.tech/renterd/api"
+	"go.sia.tech/renterd/internal/tracing"
+	"go.uber.org/zap"
+)
+
+var (
+	minBalance  = types.Siacoins(1).Div64(2).Big()
+	maxBalance  = types.Siacoins(1).Big()
+	maxNegDrift = new(big.Int).Neg(types.Siacoins(10).Big())
+)
+
+type accounts struct {
+	logger *zap.SugaredLogger
+	b      Bus
+	w      Worker
+
+	refillInterval time.Duration
+
+	mu                sync.Mutex
+	fundingContracts  []api.ContractMetadata
+	inProgressRefills map[types.PublicKey]struct{}
+}
+
+func newAccounts(l *zap.SugaredLogger, b Bus, w Worker, interval time.Duration) *accounts {
+	return &accounts{
+		b:                 b,
+		inProgressRefills: make(map[types.PublicKey]struct{}),
+		logger:            l.Named("accounts"),
+		refillInterval:    interval,
+		w:                 w,
+	}
+}
+
+func (a *accounts) markRefillInProgress(host types.PublicKey) bool {
+	a.mu.Lock()
+	defer a.mu.Unlock()
+	_, inProgress := a.inProgressRefills[host]
+	if inProgress {
+		return false
+	}
+	a.inProgressRefills[host] = struct{}{}
+	return true
+}
+
+func (a *accounts) markRefillDone(host types.PublicKey) {
+	a.mu.Lock()
+	defer a.mu.Unlock()
+	_, inProgress := a.inProgressRefills[host]
+	if !inProgress {
+		panic("releasing a refill that hasn't been in progress")
+	}
+	delete(a.inProgressRefills, host)
+}
+
+func (a *accounts) UpdateContracts(ctx context.Context, cfg api.AutopilotConfig) {
+	contracts, err := a.b.Contracts(ctx, cfg.Contracts.Set)
+	if err != nil {
+		a.logger.Errorw(fmt.Sprintf("failed to fetch contract set for refill: %v", err))
+		return
+	}
+	a.mu.Lock()
+	defer a.mu.Unlock()
+	a.fundingContracts = append(a.fundingContracts[:0], contracts...)
+}
+
+func (a *accounts) refillWorkersAccountsLoop(stopChan <-chan struct{}) {
+	ticker := time.NewTicker(a.refillInterval)
+
+	for {
+		select {
+		case <-stopChan:
+			return // shutdown
+		case <-ticker.C:
+		}
+
+		a.refillWorkerAccounts()
+	}
+}
+
+// refillWorkerAccounts refills all accounts on a worker that require a refill.
+// To avoid slow hosts preventing refills for fast hosts, a separate goroutine
+// is used for every host. If a slow host's account is still being refilled by a
+// goroutine from a previous call, refillWorkerAccounts will skip that account
+// until the previously launched goroutine returns.
+func (a *accounts) refillWorkerAccounts() {
+	ctx, span := tracing.Tracer.Start(context.Background(), "refillWorkerAccounts")
+	defer span.End()
+
+	// Map hosts to contracts to use for funding.
+	a.mu.Lock()
+	contractForHost := make(map[types.PublicKey]api.ContractMetadata, len(a.fundingContracts))
+	for _, c := range a.fundingContracts {
+		contractForHost[c.HostKey] = c
+	}
+	a.mu.Unlock()
+
+	// Fund an account for every contract we have.
+	for _, contract := range contractForHost {
+		// Only launch a refill goroutine if no refill is in progress.
+		if !a.markRefillInProgress(contract.HostKey) {
+			continue // refill already in progress
```

Comment:
should we keep track of `time.Since(refillScheduled)` here or something? This feels like one of those very subtle places where accounts could build up in the `in-progress` queue and get stuck there without us ever knowing

Reply:
hm actually nevermind, it's a non-issue